overlord/servicestate/quota_control_test.go

```diff
@@ -0,0 +1,531 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate_test
+
+import (
+	"fmt"
+	"path/filepath"
+
+	. "gopkg.in/check.v1"
+
+	"github.com/snapcore/snapd/dirs"
+	_ "github.com/snapcore/snapd/overlord/devicestate"
+	"github.com/snapcore/snapd/overlord/state"
+	_ "github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/testutil"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/overlord/servicestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/snap/snaptest"
+)
+
+type quotaControlSuite struct {
+	baseServiceMgrTestSuite
+}
+
+var _ = Suite(&quotaControlSuite{})
+
+func (s *quotaControlSuite) SetUpTest(c *C) {
+	s.baseServiceMgrTestSuite.SetUpTest(c)
+
+	// we don't need the EnsureSnapServices ensure loop to run by default
+	servicestate.MockEnsuredSnapServices(s.mgr, true)
+}
+
+type quotaGroupState struct {
+	MemoryLimit quantity.Size
+	SubGroups   []string
+	ParentGroup string
+	Snaps       []string
+}
+
+func checkQuotaState(c *C, st *state.State, exp map[string]quotaGroupState) {
+	st.Lock()
+	defer st.Unlock()
+	m, err := servicestate.AllQuotas(st)
+	c.Assert(err, IsNil)
+	c.Assert(m, HasLen, len(exp))
+	for name, grp := range m {
+		expGrp, ok := exp[name]
+		c.Assert(ok, Equals, true, Commentf("unexpected group %q in state", name))
+		c.Assert(grp.MemoryLimit, Equals, expGrp.MemoryLimit)
+		c.Assert(grp.ParentGroup, Equals, expGrp.ParentGroup)
+
+		c.Assert(grp.Snaps, HasLen, len(expGrp.Snaps))
+		if len(expGrp.Snaps) != 0 {
+			c.Assert(grp.Snaps, DeepEquals, expGrp.Snaps)
+
+			// also check on the service file states
+			for _, sn := range expGrp.Snaps {
+				// meh assume all services are named svc1
+				sliceName := name
+				if grp.ParentGroup != "" {
+					sliceName = grp.ParentGroup + "-" + name
+				}
+				checkSvcAndSliceState(c, sn+".svc1", sliceName, grp.MemoryLimit)
+			}
+		}
+
+		c.Assert(grp.SubGroups, HasLen, len(expGrp.SubGroups))
+		if len(expGrp.SubGroups) != 0 {
+			c.Assert(grp.SubGroups, DeepEquals, expGrp.SubGroups)
+		}
+	}
+}
+
+func checkSvcAndSliceState(c *C, snapSvc string, group string, sliceMem quantity.Size) {
```

Comment:
group here is more  dashedGroup or sliceName ?

Reply:
done in https://github.com/snapcore/snapd/pull/10216